### PR TITLE
Implement patch control points sample

### PR DESCRIPTION
**Pull request by Zbigniew-Jedlinski-Mobica**
## Description
Demonstrate how to use patch control points in extended dynamic state 2, which allows dynamic change of patch control points in tessellation.
Sample requires new model in assets: https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/21

## Notes
Sample was tested using following configurations:

Windows 11 Education 22H2
NVIDIA GeForce RTX 3080 Ti (NVIDIA driver version: 31.0.15.3118, driver date: 24.02.2023)
Intel(R) Core(TM) i7-9700K
16GB RAM

Windows 10 Pro
Nvidia T600 Laptop GPU (NVIDIA driver version: 31.0.15.2849, driver date: 02.02.2023)
Intel i7-11850H
32GB RAM

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

**Comment by SaschaWillems** [comment]
The sample works fine for me on Win10 with an NV RT 2060.

A few things I noticed during my review:

- Camera rotation on the horizontal axis is inverted (compared to all other samples)
- Tessellation should be enabled by default
- Default camera position is so far away, that tessellation is not visible until you move closer to the terrain. Default camera position should be moved closer.
- What is the difference between the two terrain patches? Both look the same, and both change their tessellation when moving the slider. Shouldn't the sample demonstrate a difference between the static and dynamic setup?

A minor note on naming: I'm not a native speaker, but afaik it should be  `dynamic_tessellation` instead of `dynamically_tessellation`. Some for `statically_tessellation`.

**Comment by Zbigniew-Jedlinski-Mobica** [comment]
> 

Camera rotation is fixed.
Tesselation is enable by default.
Camera position now is closer to terrain model.
Terrain patches have pipelines built in different ways. Static is built without using patch control points. Dynamic is built using patch control points that can be modified in the cpp file.

**Comment by Zbigniew-Jedlinski-Mobica** [comment]
@SaschaWillems , does it look fine now after the changes?

File: samples/extensions/patch_control_points/README.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+This sample demonstrates how to use `VK_EXT_extended_dynamic_state2` extension, which eliminates the need to create multiple pipelines in case of specific different parameters.
+
+This extension changes how **Patch Control Points** are managed. Instead of static description during pipeline creation, this extension allows developers to change the parameter by using a function before every draw.
+
+Below is a comparison of common Vulkan static and dynamic setting of patch control points in the  tessellation.
+  
+| Static/Non-dynamic | Dynamic State 2 |
+| ------------------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------- |
+| dynamic_state = \{} | dynamic_state = {VK_DYNAMIC_STATE_VIEWPORT,<br>VK_DYNAMIC_STATE_SCISSOR,<br>**VK_DYNAMIC_STATE_PATCH_CONTROL_POINTS_EXT**} |
```
Comment (asuessenbach) [inline samples/extensions/patch_control_points/README.md]: List either VIEWPORT and SCISSOR here on both sides or on neither.

File: samples/extensions/patch_control_points/README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	vkb::initializers::pipeline_dynamic_state_create_info(
+	    dynamic_state_enables.data(),
+	    static_cast<uint32_t>(dynamic_state_enables.size()),
+	    0);
+
+VkGraphicsPipelineCreateInfo graphics_create{VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_CREATE_INFO};
+graphics_create.pInputAssemblyState = &input_assembly_state;
+graphics_create.pDynamicState       = &dynamic_state;
+graphics_create.layout              = pipeline_layouts.dynamically_tessellation;
+...
```
Comment (asuessenbach) [inline samples/extensions/patch_control_points/README.md:105]: Maybe note, that you still have to set `graphics_create.pTessellationState` to point to some valid `VkPipelineTessellationStateCreateInfo`

File: samples/extensions/patch_control_points/patch_control_points.cpp [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+	/* Common uniform buffer */
+	ubo_common.projection = camera.matrices.perspective;
+	ubo_common.view       = camera.matrices.view;
+	uniform_buffers.common->convert_and_update(ubo_common);
+
+	/* Tessellation uniform buffer */
+	ubo_tess.tessellation_factor = gui_settings.tess_factor;
+	if (!gui_settings.tessellation)
+	{
+		// Setting this to zero sets all tessellation factors to 1.0 in the shader
```
Comment (asuessenbach) [inline samples/extensions/patch_control_points/patch_control_points.cpp]: factors -> levels

File: samples/extensions/patch_control_points/patch_control_points.cpp [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+		ubo_tess.tessellation_factor = 0.0f;
+	}
+
+	/* Dynamically tessellation */
+	uniform_buffers.dynamic_tessellation->convert_and_update(ubo_tess);
+
+	/* Statically tessellation */
+	uniform_buffers.static_tessellation->convert_and_update(ubo_tess);
```
Comment (asuessenbach) [inline samples/extensions/patch_control_points/patch_control_points.cpp:159]: Do the dynamic_tessellation and the static_tessellation UBs ever differ?
If not, why do you have two of them?

File: samples/extensions/patch_control_points/patch_control_points.cpp [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+	pipeline_layout_create_info.pushConstantRangeCount = 1;
+	pipeline_layout_create_info.pPushConstantRanges    = &push_constant_range;
+
+	VK_CHECK(vkCreatePipelineLayout(get_device().get_handle(),
+	                                &pipeline_layout_create_info,
+	                                nullptr,
+	                                &pipeline_layouts.static_tessellation));
+
+	/* Second descriptor set */
+	set_layout_bindings = {
```
Comment (asuessenbach) [inline samples/extensions/patch_control_points/patch_control_points.cpp:454]: Is there any difference to what set_layout_bindings was before?

File: shaders/patch_control_points/tess.tesc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+}
+ubo_tessellation;
+
+layout(vertices = 3) out;
+
+layout(location = 0) in vec3 inPos[];
+layout(location = 1) in vec3 inNormal[];
+
+layout(location = 0) out vec3 outPos[3];
+layout(location = 1) out vec3 outNormal[3];
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: outPos and outNormal are not used in tess.tese, so no need to pass them around.
outColor could then be moved to location = 0.

File: shaders/patch_control_points/tess.tesc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+layout(location = 2) out vec3 outColor[3];
+
+float GetTessLevel(vec4 p0, vec4 p1)
+{
+	float tesselationValue = 0.0f;
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: tessellationValue -> tessellationLevel ?

File: shaders/patch_control_points/tess.tesc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	float midDistance = 1.6f;
+	float shortDistance = 1.0f;
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: make midDistance and shortDistance const ?

File: shaders/patch_control_points/tess.tesc [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	vec4 midPoint = 0.5f * (p0 + p1);
+	
+	// Calculate vector from camera to mid point
+	vec4 vCam = ubo.view * midPoint;
+
+	// Calculate vector for camera projection
+	vec4 vCamView = ubo.projection * vCam;
+
+	// Calculate the vector length
+	float AvgDistance = length(vCamView);
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: Maybe AvgDistance -> centerDistance (and midPoint -> centerPoint) ?

File: shaders/patch_control_points/tess.tesc [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	return tesselationValue;
+}
+
+void main()
+{
+	if (ubo_tessellation.tessellationFactor > 0.0)
+	{
+		gl_TessLevelOuter[0] = GetTessLevel(gl_in[2].gl_Position, gl_in[0].gl_Position);
+		gl_TessLevelOuter[1] = GetTessLevel(gl_in[0].gl_Position, gl_in[1].gl_Position);
+		gl_TessLevelOuter[2] = GetTessLevel(gl_in[1].gl_Position, gl_in[2].gl_Position);
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: outColor is calculated three times here!
Maybe introduce some getColor() function, determining the color depending on the center of the triangle?

File: shaders/patch_control_points/tess.tesc [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+layout(vertices = 3) out;
+
+layout(location = 0) in vec3 inPos[];
+layout(location = 1) in vec3 inNormal[];
+
+layout(location = 0) out vec3 outPos[3];
+layout(location = 1) out vec3 outNormal[3];
+layout(location = 2) out vec3 outColor[3];
+
+float GetTessLevel(vec4 p0, vec4 p1)
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: GetTessLevel -> getTessLevel ?

File: shaders/patch_control_points/tess.tesc [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+layout(binding = 1) uniform UBOTessellation
+{
+	float tessellationFactor;
+}
+ubo_tessellation;
+
+layout(vertices = 3) out;
+
+layout(location = 0) in vec3 inPos[];
+layout(location = 1) in vec3 inNormal[];
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: inPos and inNormal are not used here, can be removed.

File: shaders/patch_control_points/tess.vert [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#version 450
+#define PI 3.14159
+layout(location = 0) in vec3 inPos;
+layout(location = 1) in vec3 inNormal;
+
+layout(location = 0) out vec3 outPos;
+layout(location = 1) out vec3 outNormal;
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.vert]: outPos and outNormal are not used in tess.tec, so no need to calculate and pass them around here.

File: samples/extensions/patch_control_points/patch_control_points.cpp [hunk 14]
```diff
@@ -0,0 +1,611 @@
...
+
+	/* Binding description */
+	std::vector<VkVertexInputBindingDescription> vertex_input_bindings = {
+	    vkb::initializers::vertex_input_binding_description(0, sizeof(Vertex), VK_VERTEX_INPUT_RATE_VERTEX),
+	};
+
+	/* Attribute descriptions */
+	std::vector<VkVertexInputAttributeDescription> vertex_input_attributes = {
+	    vkb::initializers::vertex_input_attribute_description(0, 0, VK_FORMAT_R32G32B32_SFLOAT, 0),                               // Position
+	    vkb::initializers::vertex_input_attribute_description(0, 1, VK_FORMAT_R32G32B32_SFLOAT, offsetof(Vertex, normal)),        // Normal
```
Comment (asuessenbach) [inline samples/extensions/patch_control_points/patch_control_points.cpp]: Normal is not used in tess.vert, so need to describe it here.

**Review by asuessenbach — COMMENTED** [review]
In a sample on dynamic usage of tessellation patch control points I would expect some dynamic usage of them. That is, maybe, two or more models with varying numbers of patch control points.
Besides that, it looks really good!
Just a couple of mostly minor issues...

File: samples/extensions/patch_control_points/README.md (hunk 1 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/README.md]: Fixed

File: samples/extensions/patch_control_points/README.md (hunk 2 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/README.md:105]: Fixed

File: samples/extensions/patch_control_points/patch_control_points.cpp (hunk 3 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/patch_control_points.cpp]: Fixed

File: samples/extensions/patch_control_points/patch_control_points.cpp (hunk 4 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/patch_control_points.cpp:159]: static_tessellation is for drawing the model without using the patch control points functionality while dynamic_tessellation uses it. Two pipelines are used here to show that the same effect can be achieved with patch control points

File: samples/extensions/patch_control_points/patch_control_points.cpp (hunk 5 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/patch_control_points.cpp:454]: The set layout is set the same for static_tessellation and dynamic_tessellation

File: shaders/patch_control_points/tess.tesc (hunk 6 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Removed

File: shaders/patch_control_points/tess.tesc (hunk 7 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Fixed

File: shaders/patch_control_points/tess.tesc (hunk 8 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Fixed

File: shaders/patch_control_points/tess.tesc (hunk 9 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Fixed

File: shaders/patch_control_points/tess.tesc (hunk 10 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Function to get color was added to tesc file

File: shaders/patch_control_points/tess.tesc (hunk 11 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Fixed

File: shaders/patch_control_points/tess.tesc (hunk 12 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Removed

File: shaders/patch_control_points/tess.vert (hunk 13 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.vert]: Removed

File: samples/extensions/patch_control_points/patch_control_points.cpp (hunk 14 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/patch_control_points.cpp]: Normal was removed from code

**Comment by Zbigniew-Jedlinski-Mobica** [comment]
@asuessenbach when I loaded models with different number of control points in one pipeline, they were displayed incorrectly.
Each pipeline has a defined shader file that supports a certain number of patch control points.
As I understand this behavior is related to the way GLSL works

File: samples/extensions/patch_control_points/patch_control_points.cpp [hunk 15]
```diff
@@ -145,11 +145,11 @@ void PatchControlPoints::update_uniform_buffers()
 	uniform_buffers.common->convert_and_update(ubo_common);
 
 	/* Tessellation uniform buffer */
-	ubo_tess.tessellation_factor = gui_settings.tess_factor;
+	ubo_tess.tessellation_level = gui_settings.tess_level;
 	if (!gui_settings.tessellation)
 	{
-		// Setting this to zero sets all tessellation factors to 1.0 in the shader
-		ubo_tess.tessellation_factor = 0.0f;
+		// Setting this to zero sets all tessellation levels to 1.0 in the shader
```
Comment (asuessenbach) [inline samples/extensions/patch_control_points/patch_control_points.cpp]: Is this comment still valid?

File: shaders/patch_control_points/tess.tesc [hunk 16]
```diff
@@ -31,58 +31,69 @@ ubo_tessellation;
...
-		tesselationValue = 1.0f;
-		outColor[gl_InvocationID] = vec3(1.0f, 0.0f, 0.0f); // red color
+		tessellationLevel = 1.0f;
 	}
-	else if (AvgDistance >= shortDistance && AvgDistance < midDistance)
+	else if (centerDistance >= shortDistance && centerDistance < midDistance)
 	{
-		tesselationValue = ubo_tessellation.tessellationFactor * 0.4;
-		outColor[gl_InvocationID] = vec3(0.0f, 0.0f, 1.0f); // blue color
+		tessellationLevel = ubo_tessellation.tessellationFactor * 0.4;
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: 0.4 -> 0.4f

File: shaders/patch_control_points/tess.tesc [hunk 17]
```diff
@@ -31,58 +31,69 @@ ubo_tessellation;
...
+}
+
+vec3 getColor(float tessellationLevel)
+{
+	vec3 outColor;
+	if (tessellationLevel == 1.0f)
+	{
+		outColor = vec3(1.0f, 0.0f, 0.0f);        // red color
+	}
+	else if (tessellationLevel == ubo_tessellation.tessellationFactor * 0.4)
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.tesc]: 0.4 -> 0.4f

File: shaders/patch_control_points/tess.vert [hunk 18]
```diff
@@ -35,16 +31,13 @@ layout(push_constant) uniform Push_Constants
...
-	rotX[1][2] = sin(PI);
-	rotX[2][1] = -sin(PI);
-	rotX[2][2] = cos(PI);
-	vec4 pos = rotX * vec4(inPos + push_constants.direction, 1.0f);
+	mat4 rotX   = mat4(1.0f);
+	rotX[1][1]  = cos(PI);
+	rotX[1][2]  = sin(PI);
+	rotX[2][1]  = -sin(PI);
+	rotX[2][2]  = cos(PI);
+	vec4 pos    = rotX * vec4(inPos + push_constants.direction, 1.0f);
```
Comment (asuessenbach) [inline shaders/patch_control_points/tess.vert]: skip `pos` and directly assign the result to `gl_Position`?

File: samples/extensions/patch_control_points/patch_control_points.cpp (hunk 5 above)
Comment (asuessenbach) [reply samples/extensions/patch_control_points/patch_control_points.cpp:454]: Why do you initialize it again, then?

**Comment by asuessenbach** [comment]
> Each pipeline has a defined shader file that supports a certain number of patch control points.

Then why use a dynamic number of patch control points?

**Comment by Zbigniew-Jedlinski-Mobica** [comment]
> Then why use a dynamic number of patch control points?

The use of a dynamic number of patch control points is to cover the use of dynamic state 2. Dynamic state 1, 2 and 3 provide the functionality of using a dynamic pipeline instead of a static one.

File: samples/extensions/patch_control_points/patch_control_points.cpp (hunk 5 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/patch_control_points.cpp:454]: > Why do you initialize it again, then?

There are two initializations to make it more visible that one is used for the static pipeline and the other for the dynamic pipeline.

File: samples/extensions/patch_control_points/patch_control_points.cpp (hunk 15 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply samples/extensions/patch_control_points/patch_control_points.cpp]: Fixed

File: shaders/patch_control_points/tess.tesc (hunk 16 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Fixed

File: shaders/patch_control_points/tess.tesc (hunk 17 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.tesc]: Fixed

File: shaders/patch_control_points/tess.vert (hunk 18 above)
Comment (Zbigniew-Jedlinski-Mobica) [reply shaders/patch_control_points/tess.vert]: Fixed

**Comment by Zbigniew-Jedlinski-Mobica** [comment]
> Why do you initialize it again, then?

There are two initializations to make it more visible that one is used for the static pipeline and the other for the dynamic pipeline.

**Review by SaschaWillems — APPROVED** [review]

**Review by asuessenbach — APPROVED** [review]

**Comment by JoseEmilio-ARM** [comment]
Tested on Android (Pixel 6), where extension is supported and sample seems to run fine, but the results do not look as expected e.g. on Windows. Maybe it is just wireframe not working as expected, but also increasing tessellation level does not have an effect. 

Screenshot_20230703-150544

Code looks good however, approving. Android not a blocker as this extension might have limited use on mobile.

**Review by JoseEmilio-ARM — APPROVED** [review]

**Pull request merged**